tutorials/math/combinatorics.md

```diff
@@ -1,10 +1,215 @@
 ---
 title: 'Combinatorics'
-description: ''
-hide_table_of_contents: true
+description: 'nCr and its applications'
+hide_table_of_contents: false
 draft: true
 keywords:
   - leetcode
   - tutorial
   - combinatorics
----
\ No newline at end of file
+---
+
+<TutorialAuthors names="@BlackPanther112358"/>
+
+## Overview
+
+Combinatorics is primarily involved with art of counting. This includes counting no. of ways for a certain position to occur, to arrange some objects according to given rules or choosing some objects from a collection. Quite often, the required number takes a very large value, thus it is a very common practice to return the answer by taking modulo with some prime number $p$ (which is $1e9 + 7$ quite often, and quite recently $998244353$ also has become prominent). You can read more about modulo [here](../../tutorials/basic-topics/mod.md)
+
+## Finding $n \choose r$
+
+Pronounced "n choose r", this is the mathematical notation to represent number of ways to choose r objects out of a collection of n objects. The analytical formula can be written as $n \choose r$ $=$ $\frac{n!}{r!\,(n - r)!}$.
+
+This leads to a neat recurrence relation :-  $n \choose r$ $=$ $n - 1 \choose r$ + $n - 1 \choose r - 1$ $\\$
+We can precompute all the required values using the above formula in $O(n^2)$ and then perform lookup in $O(1)$ time. The resulting values also form [Pascal's Triangle](https://en.wikipedia.org/wiki/Pascal%27s_triangle).
+
+### Example #1 [2221 - Find Triangular Sum of an Array](https://leetcode.com/problems/find-triangular-sum-of-an-array/)
+
+The important insight here is that the figure provided is nothing but an inverted Pascal's Triangle and contribution of each cell in the final sum is $value\,of\,cell$ $*$ $binomial\,coefficient\,at\,the\,particular\,position\,in\,Pascal's\,Triangle$
+
+Thus for the cell at $i^{th}$ index in the topmost row, it's value is multiplied by $n - 1 \choose i$ and added to the final sum $modulo\,10$.
+Time Complexity of the program is $O(n^2)$ for computing the binomial coefficient and $O(n)$ Space complexity.
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+
+```cpp
+class Solution {
+public:
+    int triangularSum(vector<int>& nums) {
+        int n = nums.size();
+        vector<int> pascalTriangleRow = {1};
+        // Calculate the ith row using (i - 1)th row
+        for(int i = 0; i < n; i++){
+            vector<int> nextRow = {1};
+            for(int j = 1; j < i; j++){
+                nextRow.push_back((pascalTriangleRow[j] + pascalTriangleRow[j - 1])%10);
```

Comment:
add space before and after an operator. apply to all

```
nextRow.push_back((pascalTriangleRow[j] + pascalTriangleRow[j - 1]) % 10);
```

Reply:
ok